docs/rfcs/041-sharded-ingest.md

```diff
@@ -0,0 +1,267 @@
+
+### Tweaking the Pagserver Safekeeper Selection Algorithm
+
+We could make the pageserver aware of which safekeeper's already host shards for the timeline along
+with their current WAL positions. The pageserver should then prefer safkeepers that are in the same
+AZ _and_ already have a shard with a position close to the desired start position.
+
+We currently run one safekeeper per AZ, so the point is mute until that changes.
+
+### Pipelining first ingest phase
+
+The first ingest phase is a stateless transformation of a binary WAL record into a pre-processed
+output per shard.  To put multiple CPUs to work, we may pipeline this processing up to some defined buffer
+depth.
+
+### Perhaps: Serve reads from the safekeeper/eliminate pageserver InMemoryLayer
+
+Now that we are doing some pre-processing of the WAL on the safekeeper and storing
+a buffer of some recent reads in Cursor, it would be possible to serve some reads
+from the safekeeper directly.  It's unclear whether this is a desirable change: it would
+only be a net benefit if it enabled us to avoid doing some work on the pageserver.
+
+Ultimately, if the safekeeper served recent reads, we could build entire L0 layers on the safekeeper
+and upload them to S3, saving the pageserver the disk bandwidth of writing ephemeral layers
+and then flushing them to L0s, since ephemeral layers are approximately just extents of WAL, post-processed
+into a readable form.
+
+## Alternatives considered
+
+### Give safekeepers enough state to fully decode WAL
+
+In this RFC, we only do the first phase of ingest on the safekeeper, because this is
+the phase that is stateless.  Subsequent changes then happen on the pageserver, with
+access to the `Timeline` state.
+
+We could do more work on the safekeeper if we transmitted metadata state to the safekeeper
+when subscribing to the WAL: for example, by telling the safekeeper all the relation sizes,
+so that it could then generate all the metadata writes for relation sizes.
+
+We avoid doing this for several reasons:
+1. Complexity: it's a more invasive protocol change
+2. Decoupling: having the safekeeper understand the `ProcessedWalIngest` already somewhat
+   infects it with knowledge of the pageserver, but this is mainly an abstract structure
+   that describes postgres writes.  However, if we taught the safekeeper about the exact
+   way that pageserver deals with metadata keys, this would be a much tighter coupling.
+3. Load: once the WAL has been processed to the point that it can be split between shards,
+   it is preferable to share out work on the remaining shards rather than adding extra CPU
+   load to the safekeeper.
+
+### Do pre-processing on the compute instead of the safekeeper
+
+Since our first stage of ingest is stateless, it could be done at any stage in the pipeline,
+all the way up to the compute.
+
+We choose not to do this, because it is useful for the safekeeper to store the raw WAL rather
+than just the preprocessed WAL:
```

Comment:
I would also add that the raw WAL needs to have a stable protocol, since we may want to reingest it at an arbitrary point in the future, while we may want to change the interpreted protocol for more efficient Pageserver processing -- storing the raw WAL gives us much more flexibility here in the future.